_posts/2022-11-02-#26140.md

```diff
@@ -0,0 +1,62 @@
+---
+layout: pr
+date: 2022-11-02
+title: "Move CNodeState members guarded by g_msgproc_mutex to Peer"
+pr: 26140
+authors: [dergoegge]
+components: ["p2p", "refactoring"]
+host: dergoegge
+status: upcoming
+commit:
+---
+
+## Notes
+
+- [#26036](https://github.com/bitcoin/bitcoin/pull/26036) introduced a new
+  mutex (`NetEventsInterface::g_msgproc_mutex`) to document the fact that our
+  message processing code is single threaded (`msghand` thread). Any
+  `PeerManagerImpl` or `Peer` members that are only ever accessed from that
+  single thread should be annotated as `GUARDED_BY(g_msgproc_mutex)`, to avoid
+  bugs where those members are accessed by other threads as well (in which case
+  they need to be guarded by a different mutex).
+
+- `CNodeState` is documented as only containing validation specific state and
+  therefore has all its members guarded by `cs_main`. In reality, not all
+  `CNodeState` are validation specific and should be moved from `CNodeState` to
+  `Peer`.
+
+- [#26140](https://github.com/bitcoin/bitcoin/pull/26140) is a simple refactor
+  PR that moves some of the non-validation relevant `CNodeState` members to
+  `Peer` and annotates them as guarded by `g_msgproc_mutex`.
+
+## Questions
+
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+   NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
+
+2. What is `cs_main` and what is it used for?
+
+3. Which threads access state relevant for message processing
+   (`PeerManagerImpl`, `Peer`, etc.)? (Hint: try grepping for `std::thread` to
+   find out which threads exist)
```

Comment:
Not sure if you want to add it, but there's also this overview in the developer notes. But grepping might be more fun.

Reply:
Yea the dev docs are better.